JustLog/Classes/LogstashDestination.swift

```diff
@@ -8,30 +8,46 @@
 
 import Foundation
 import SwiftyBeaver
-import CocoaAsyncSocket
 
+/// This entire class relies on `logDispatchQueue` to synchronise access to the `logsToShip` dictionary
+/// Every action is put in the `logDispatchQueue` which is synchronous, even though the actual sending
+/// via `URLSession` is not.
+/// Here's a brief summary of what operations are added to the queue and when:
+/// A log is generated via `send` and added to the `logsToShip` dictionary (along with a `tag` identifier).
+/// At some point `writeLogs` is called that creates an `NSURLSessionStreamTask` to send the log to the server.
+/// When the task completes, either `completeLog` or `retryLog` are called.
+///    `completeLog` removes the log from `logsToShip`
+///    `retryLog` clears the log from `logsInSocketQueue` and causes it to be resent the next time `writeLogs` is called.
+/// An optional `completionHandler` is called when all logs are written.
 public class LogstashDestination: BaseDestination  {
     
+    /// Settings
+    var logActivity: Bool = false
     public var logzioToken: String?
     
+    /// Logs buffer
     var logsToShip = [Int : [String : Any]]()
-    fileprivate var completionHandler: ((_ error: Error?) -> Void)?
+    let logDispatchQueue = OperationQueue()
+    
+    /// Private
+    private var completionHandler: ((_ error: Error?) -> Void)?
     private let logzioTokenKey = "token"
     
-    var logActivity: Bool = false
-    let logDispatchQueue = OperationQueue()
-    var socketManager: AsyncSocketManager!
+    /// Socket
+    private let socket: LogstashDestinationSocketProtocol
+    private var logsInSocketQueue = [Int]()
     
     @available(*, unavailable)
     override init() {
         fatalError()
     }
     
-    public required init(host: String, port: UInt16, timeout: TimeInterval, logActivity: Bool, allowUntrustedServer: Bool = false) {
+    public required init(socket: LogstashDestinationSocketProtocol, logActivity: Bool) {
+        
+        self.socket = socket
         super.init()
         self.logActivity = logActivity
         self.logDispatchQueue.maxConcurrentOperationCount = 1
```

Comment:
It will be easier to understand if the set of the max operation will be done in the same place of the queue allocation